tactics/class_tactics.ml

```diff
@@ -428,19 +419,38 @@ and e_my_find_search db_list local_db secvars hdc complete only_classes env sigm
         match repr_hint t with
         | Extern _ -> (tac, b, true, name, lazy (pr_hint env sigma t ++ pp))
         | _ -> (tac, b, false, name, lazy (pr_hint env sigma t ++ pp))
-  in List.map tac_of_hint hintl
+  in
+  let hint_of_db = hintmap_of sigma hdc secvars concl in
+  let hintl = List.map (fun db -> db, hint_of_db db) (local_db :: db_list) in
+  (* In case there is a mode mismatch in all the databases we get stuck.
+     Otherwise we consider the hints that match.
+     Recall the local database uses the union of all the modes in the other databases. *)
+  if List.for_all (function (db, ModeMatch _) -> false | (db, ModeMismatch) -> true) hintl then
+    ModeMismatch
+  else
+    let hintl = List.map_filter (function (db, ModeMatch l) -> Some (db, l) | (db, ModeMismatch) -> None) hintl in
```

Comment:
If we merge the list operations here I think it's clearer:
```suggestion
  let hintl = List.map_filter (fun db -> match hint_of_db db with
      | ModeMatch l -> Some (db, l)
      | ModeMismatch -> None)
      (local_db :: db_list)
  in
  (* In case there is a mode mismatch in all the databases we get stuck.
     Otherwise we consider the hints that match.
     Recall the local database uses the union of all the modes in the other databases. *)
  if List.is_empty hintl then ModeMismatch
  else
```